### ISSUE TITLE
Ingress Resource - Strip Prefix from URL for path based routing

### ISSUE THREAD

**Issue by neeseius** [F5Networks/k8s-bigip-ctlr#835]
#### Description
It would be very useful if there were an annotation for, by default, stripping the path-prefix from paths.
example:
```
  - host: apitesting.example.com
    http:
      - path: /testenv3/api1
        backend:
          serviceName: testenv3-api1
          servicePort: 443
      - path: /testenv1/api1
        backend:
          serviceName: testenv1-api1
          servicePort: 443
```
So paths would be modified like below when they hit the backend service:
https://apitesting.example.com/testenv3/api1/check/token
becomes:
/check/token
instead of:
/testenv3/api1/check/token

This is useful because the backend service may not understand the paths being defined in the Ingress. We can then utilize paths defined in the Ingress to do a lot more and act as an api-gateway for example.

**Comment by neilcowlin** [comment]
+1, I'm blocked needing this functionality (basepath stripping) too.  Apps shouldn't have to be aware of how they've been exposed.  It's really limiting to portability and flexibility.

#621 was asking for similar.  It was closed with the addition of some rewrite features, but I don't think the actual issue/enhancement was addressed - or at least I haven't been able to figure it out how to make it work!

Neither of the two rewrite options we have are suitable -

[rewrite-app-root](https://clouddocs.f5.com/containers/v2/kubernetes/kctlr-rewrite-url.html#rewrite-app-root)
_Redirect "/" to "/home" for all traffic_
i.e. it's just a 302 for the browser

[rewrite-target-url](https://clouddocs.f5.com/containers/v2/kubernetes/kctlr-rewrite-url.html#rewrite-target-url)
I can get this to rewrite basepaths to / for multiple services, but _only_ those paths - they are truncated.
i.e. with
`virtual-server.f5.com/rewrite-target-url: "rewrite.bar.com/app1/=rewrite.bar.com/,rewrite.bar.com/app2=rewrite.bar.com/" `
_rewrite.bar.com/app1/real-path_ rewrites to _rewrite.bar.com/_ (note no /real-path)
It's close!

The behaviour I think we're missing is as per [ingress-nginx](https://github.com/kubernetes/ingress-nginx/tree/master/docs/examples/rewrite)
_rewrite.bar.com/app1/real-path_ rewrites to _rewrite.bar.com/real-path_

A neat, simple, even boolean annotation to achieve this would be amazing.

**Comment by mdditt2000** [comment]
@neilcowlin @neeseius this feature will be available in early April using a AS3 declaration in the cfgmap. AS3 support a full rewrite/redirect module for HTTP Header, uri etc. Read more about AS3 at https://clouddocs.f5.com/products/extensions/f5-appsvcs-extension/latest/declarations/http-services.html#using-a-local-traffic-policy-to-forward-http-requests

At this time we will not be adding additional annotations.

**Comment by neilcowlin** [comment]
Sounds good - ctlr+as3 should be a powerful combo.  Can't wait to give that a try.

**Comment by mdditt2000** [comment]
@neilcowlin we are on-track for April 10th RTW. I would recommend start creating looking at AS3 and creating the declaration. There is alot of content already created to help out. CIS 1.9 will consume your AS3 declaration.

**Comment by neeseius** [comment]
Are there any docs/tutorials coming soon on how to use as3 declarations in configmaps. And can we have multiple AS3 configmaps?

**Comment by rsalsa** [comment]
> Are there any docs/tutorials coming soon on how to use as3 declarations in configmaps. And can we have multiple AS3 configmaps?

In this release, CIS supports only one AS3 ConfigMap instance. The documentation is located on our  [Clouddocs](https://clouddocs.f5.com/containers/v2/kubernetes/kctlr-k8s-as3.html) site.

**Comment by mdditt2000** [comment]
There will always be one AS3 instance since it is a declarative API. The AS3 declaration can include many applications or services. Here is a sample of three applications. http,https, waf https://github.com/mdditt2000/kubernetes/blob/dev/cis-1-9/A3/f5-as3-configmap.yaml -- Use the kubectl apply -f filename to add or remove new applications. There are 65 AS3 example at https://github.com/F5Networks/f5-appsvcs-extension/blob/master/dist/latest/as3-examples-3.10.0-5.collection.json - Import the json into postman

**Comment by mdditt2000** [comment]
@neeseius can you reach out to me directly so i can help build some AS3 configmaps.

**Comment by ramnes** [comment]
The behavior OP describes should either be the default or have an easy way to be configured, just like all the other ingress-controllers out there. AS3 seems very inconvenient for such a basic use-case, and it's actually having us thinking about going back to ingress-nginx. Please reconsider your position!

**Comment by mdditt2000** [comment]
PM CONTCNTR-1160 for tracking

**Comment by mdditt2000** [comment]
CONTCNTR-1160 is still backlogged.

**Comment by mdditt2000** [comment]
No plan to add this feature to Ingress. Please use CRD released in CIS 2.2. Here are the examples https://github.com/F5Networks/k8s-bigip-ctlr/blob/master/docs/_static/config_examples/crd/PathRewrite/README.md

Closing issue
